mongo/collection.go

```diff
@@ -1664,7 +1664,94 @@ func (coll *Collection) Indexes() IndexView {
 
 // Drop drops the collection on the server. This method ignores "namespace not found" errors so it is safe to drop
 // a collection that does not exist on the server.
-func (coll *Collection) Drop(ctx context.Context) error {
+func (coll *Collection) Drop(ctx context.Context, opts ...*options.DropCollectionOptions) error {
+	dco := options.MergeDropCollectionOptions(opts...)
+	efc := dco.EncryptedFields
+	var err error
+	if efc == nil {
+		efc = coll.db.getEncryptedFieldsFromMap(coll.name)
+	}
+	if efc == nil && coll.db.client.encryptedFieldsMap != nil {
+		efc, err = coll.db.getEncryptedFieldsFromServer(ctx, coll.name)
+		if err != nil {
+			return err
+		}
+	}
+
+	if efc != nil {
+		return coll.dropEncryptedCollection(ctx, efc)
+	}
+
+	return coll.drop(ctx)
+}
+
+// dropEncryptedCollection drops collection with EncryptedFields.
+func (coll *Collection) dropEncryptedCollection(ctx context.Context, efc interface{}) error {
+	var efcBSON bsoncore.Document
+	efcBSON, err := transformBsoncoreDocument(coll.registry, efc, true /* mapAllowed */, "encryptedFields")
+	if err != nil {
+		return fmt.Errorf("error in MarshalWithRegistry: %v", err)
+	}
+
+	// Drop the data collection.
+	err = coll.drop(ctx)
+	if err != nil {
+		return err
+	}
+	// Drop the state collections ESCCollection, ECCCollection, and ECOCCollection.
+	// Drop ESCCollection.
+	escCollection := "enxcol_." + coll.name + ".esc"
```

Comment:
The code for checking for the presence of `*Collection` in `efcBSON`, dropping that string if it exists, and using `enxcol_.<collname>.*` if the value doesn't exist is a little hard to read IMO.

I might do something like:
```
escCollName := "enxcol_." + coll.name + ".esc"
if val, err := efcBSON.LookupErr("escCollection"); err == nil {
   var ok bool
   if escCollName, ok = val.StringValueOK(); !ok {
      return fmt.Errorf("expected type string for 'escCollection', got: %v", val.Type)
   }
}

```
I _think_ errors beyond `bsoncore.ErrElementNotFound` due to a malformed document would have been found in the previous call to `transformBsoncoreDocument`.